File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
// import java.util.*;
class Solution {
    boolean[] vis;
    int[][] net;
    int N;
    int answer = Integer.MAX_VALUE;
    
    public int solution(int n, int[][] wires) {
        N = n;
        vis = new boolean[n];
        net = new int[n][n];
        for(int[] wire : wires){
            net[wire[0]-1][wire[1]-1] = 1;
            net[wire[1]-1][wire[0]-1] = 1;
        }
        dfs(1);
        return answer;
    }
    public int dfs(int i){
        vis[i] = true;
        int child = 1;
        for(int j = 0; j<N; j++){
            if(net[i][j]==1&&!vis[j]){
                child += dfs(j);
            }
        }
        answer = Math.min(answer, Math.abs(N-2*child));
        return child;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
using namespace std;

bool vis[101];
int net[101][101];
int answer = 101;
int dfs(int i, int n){
    
    vis[i] = true;
    int child = 1;
    for(int j = 1; j<=n; j++){
        if(net[i][j]==1&&!vis[j]){
            child+= dfs(j,n);
        }
    }
    
    answer = min(answer, abs(n-child*2));
    return child;
}
int solution(int n, vector<vector<int>> wires) {
    for(vector<int> v : wires){
        net[v[0]][v[1]] = 1;
        net[v[1]][v[0]] = 1;
    }
    dfs(1,n);
    
    return answer;
}
